### ISSUE TITLE
Performance optimization

### ISSUE THREAD

**Issue by yufeih** [dotnet/docfx#2972]
Build _azure-docs-pr_ currently takes more than 6 min, investigate and optimize build time.

- [ ] Do not load html twice to calculate word count
- [ ] CPU usage is around 20%, it should be higher

**Comment by yufeih** [comment]
Creating a new `MarkdownPipeline` allocates lots of memory, especially in the [EmojiParser](https://github.com/lunet-io/markdig/blob/master/src/Markdig/Extensions/Emoji/EmojiParser.cs#L67). These tiny allocations literally turns our parallel processing into serial processing.

**Concurrency Visualizer** shows most threads are just sit there waiting (red part)

![image](https://user-images.githubusercontent.com/511355/42299430-b03b8f3c-803d-11e8-9ffc-a9856eba4276.png)

**Resource contention call tree** shows EmojiParser is heavily contended
![image](https://user-images.githubusercontent.com/511355/42299284-f644ab18-803c-11e8-9e8a-577db3676fb7.png)


**Functions view** shows the most contended functions are dictionary insersion and allocation (*JIT_New*)
![image](https://user-images.githubusercontent.com/511355/42299401-882d0a20-803d-11e8-841c-e407bc658502.png)
![image](https://user-images.githubusercontent.com/511355/42299511-24ced412-803e-11e8-874d-78ce59dfbbac.png)

The fix is reusing a _single_, _global_  `MarkdownPipeline` instance for all contents.

@herohua @qinezh @yishengjin1413 @OsmondJiang @superyyrrzz @live1206 @928PJY 

cc @xoofx

**Comment by xoofx** [comment]
Yes, a MarkdownPipeline should be cached, it can be heavy to initialized and was actually built immutable so that multiple concurrent parsers can work with the same pipeline. Though, the TextMatchHelper class will require an optimization pass at some point (there is a TODO in the code about this memory consumption problem)

**Comment by yufeih** [comment]
@xoofx , some pipelines are not thread safe. [this](https://github.com/lunet-io/markdig/blob/master/src/Markdig/Extensions/AutoIdentifiers/AutoIdentifierExtension.cs#L35) is an example I found so far, `headingWriter` is shared in `HeadingBlock_ProcessInlinesEnd`

**Comment by xoofx** [comment]
> @xoofx , some pipelines are not thread safe. this is an example I found so far, headingWriter is shared in HeadingBlock_ProcessInlinesEnd

Yeah, so it's an oversight when I was developing all the extensions, I might have leak some non thread safe extensions, so that in that case, we should fix them. Please log an issue on markdig and if you have more, I would be glad to have some help to get all over them, thanks!
